README.md
# Terraform Provider for Cohesity

- Website: https://www.terraform.io
- [![Gitter chat](https://badges.gitter.im/hashicorp-terraform/Lobby.png)](https://gitter.im/hashicorp-terraform/Lobby)
- Mailing list: [Google Groups](http://groups.google.com/group/terraform-tool)

<img src="https://cdn.rawgit.com/hashicorp/terraform-website/master/content/source/assets/images/logo-hashicorp.svg" width="600px">

## Maintainers

This provider plugin is maintained by [Cohesity](https://www.cohesity.com/)

## Table of contents :scroll:

 - [Requirements](#requirements)
 - [Building the Provider](#building-the-provider)
 - [Using the Provider](#using-the-provider)
 - [Developing the Provider](#developing-the-provider)
 - [How can you contribute](#contribute)
 - [Suggestions and Feedback](#suggest)


## <a name ="requirements"></a> Requirements :clipboard:

-	[Terraform](https://www.terraform.io/downloads.html) 0.12.20+
-	[Go](https://golang.org/doc/install) 1.12 (to build the provider plugin)
-   [Cohesity](https://www.cohesity.com/) DataPlatform 6.4+

## <a name ="building-the-provider"></a> Building The Provider :pencil2:

Clone repository to: `$GOPATH/src/github.com/terraform-providers/terraform-provider-cohesity`

```sh
$ git clone git@github.com:terraform-providers/terraform-provider-cohesity $GOPATH/src/github.com/terraform-providers/terraform-provider-cohesity
```

Enter the provider directory and build the provider

```sh
$ cd $GOPATH/src/github.com/terraform-providers/terraform-provider-cohesity
$ make build
```

The provider binary can be found in `$GOPATH/bin` directory.

## <a name ="using-the-provider"></a> Using the provider :arrow_forward:

The Cohesity provider documentation can be found on [provider's website](https://www.terraform.io/docs/providers/cohesity/index.html)

## <a name ="developing-the-provider"></a> Developing the Provider :hammer_and_pick:

If you wish to work on the provider, you'll first need [Go](http://www.golang.org) installed on your machine (version 1.12 is *required*). You'll also need to correctly setup a [GOPATH](http://golang.org/doc/code.html#GOPATH), as well as adding `$GOPATH/bin` to your `$PATH`.

To compile the provider, run `make build`. This will build the provider and put the provider binary in the `$GOPATH/bin` directory.

```sh
$ make build
...
$ $GOPATH/bin/terraform-provider-cohesity
...
```
In order to test the provider, you can simply run `make test`.

```sh
$ make test
```
In order to run a specific acceptance test, change the variables in acceptance_test_variables.go to suit your environment and run `make testacc`

*Note:* Set the environment variables for configuring the Cohesity provider and sensitive resource arguments not set in the acceptance_test_variables.go file

```sh
$ make testacc TEST=./cohesity TESTARGS='-run=TestAccVirtualEditionCluster_basic'
```

## <a name="contribute"></a> Contribute :handshake:

* [Refer our contribution guideline](./CONTRIBUTING.md).


## <a name ="suggest"></a> Suggestions or Feedback :raised_hand:

We would love to hear from you. Please send your questions and feedback to: *cohesity-api-sdks@cohesity.com*


CHANGELOG.md
## 1.1.0 (Unreleased)
## 1.0.0 (July 07, 2020)

FEATURES:

* **New Resource:** `cohesity_source_vmware`
* **New Resource:** `cohesity_job_vmware`
* **New Resource:** `cohesity_job_run`
* **New Resource:** `cohesity_job_restore_vmware_vm`

## 0.1.0 (June 02, 2020)

FEATURES:

* **New Resource:** `cohesity_virtual_edition_cluster`
* **New Resource:** `cohesity_cloud_edition_cluster`
* **New Resource:** `cohesity_physical_edition_cluster`


cohesity/provider_test.go
package cohesity

import (
	"testing"

	"github.com/hashicorp/terraform/helper/schema"
	"github.com/hashicorp/terraform/terraform"
)

var testAccProviders map[string]terraform.ResourceProvider
var testAccProvider *schema.Provider

func init() {
	testAccProvider = Provider().(*schema.Provider)
	testAccProviders = map[string]terraform.ResourceProvider{
		"cohesity": testAccProvider,
	}
}

func TestProvider(t *testing.T) {
	if err := Provider().(*schema.Provider).InternalValidate(); err != nil {
		t.Fatalf("err: %s", err)
	}
}


cohesity/resource_cohesity_job_run_test.go
package cohesity

import (
	"errors"
	"fmt"
	"log"
	"testing"

	CohesityManagementSdk "github.com/cohesity/management-sdk-go/managementsdk"
	"github.com/hashicorp/terraform/helper/resource"
	"github.com/hashicorp/terraform/terraform"
)

func TestAccJobRun_basic(t *testing.T) {
	resource.Test(t, resource.TestCase{
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckJobRunDestroy,
		Steps: []resource.TestStep{
			{
				Config: fmt.Sprintf(testAccJobRunCreateConfig, jobVMwareSourceEndpoint, vmWareJobName, vmWareJobProtectVM),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckJobRunCreated(),
				),
			},
		},
	})
}

func testAccCheckJobRunDestroy(s *terraform.State) error {
	return nil
}

func testAccCheckJobRunCreated() resource.TestCheckFunc {
	return func(s *terraform.State) error {
		var cohesityConfig = testAccProvider.Meta().(Config)
		client, err := CohesityManagementSdk.NewCohesitySdkClient(cohesityConfig.clusterVip,
			cohesityConfig.clusterUsername, cohesityConfig.clusterPassword, cohesityConfig.clusterDomain)
		if err != nil {
			log.Printf(err.Error())
			return errors.New("Failed to authenticate with Cohesity")
		}
		jobName := vmWareJobName
		//get the protection job details
		log.Printf("[INFO] Get the protection job (%s) details", jobName)
		protectionJob, err := client.ProtectionJobs().GetProtectionJobs(nil, []string{jobName}, nil,
			nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
		if err != nil {
			log.Printf(err.Error())
			return errors.New("Failed to find the protection job")
		}
		jobID := *protectionJob[0].Id
		//get the protection runs to check the the status of recent job run
		log.Printf("[INFO] Protection job %s found with ID %d ", jobName, jobID)
		return nil
	}
}

const testAccJobRunCreateConfig = `
resource "cohesity_source_vmware" "terraform_vmware_source"{
	endpoint = "%s"
	username = "administrator"
	vmware_type = "VCenter"
	cap_streams_per_datastore = true
	number_of_streams = 5
	enable_latency_throttling = true
	new_task_latency = 110
	active_task_latency = 120
}
resource "cohesity_job_vmware" "terraform_vmware_protection_job" {
	name = "%s"
	protection_source = cohesity_source_vmware.terraform_vmware_source.endpoint
	include = ["%s"]
	policy = "Bronze"
	storage_domain = "DefaultStorageDomain"
	delete_snapshots = true
	full_sla = 200
	incremental_sla = 140
	qos_type = "BackupSSD"
	priority = "Low"
}

resource "cohesity_job_run" "terraform_vmware_job_run"{
	name = cohesity_job_vmware.terraform_vmware_protection_job.name
	timestamp = "18000000"
}`


cohesity/resource_cohesity_job_vmware_test.go
package cohesity

import (
	"errors"
	"fmt"
	"log"
	"strconv"
	"testing"

	CohesityManagementSdk "github.com/cohesity/management-sdk-go/managementsdk"
	"github.com/hashicorp/terraform/helper/resource"
	"github.com/hashicorp/terraform/terraform"
)

func TestAccJobVmware_basic(t *testing.T) {
	resource.Test(t, resource.TestCase{
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckJobVmwareDestroy,
		Steps: []resource.TestStep{
			{
				Config: fmt.Sprintf(testAccJobVmwareCreateConfig, jobVMwareSourceEndpoint, vmWareJobName),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckJobVmwareCreated(),
				),
			},
			{
				Config: fmt.Sprintf(testAccJobVmwareUpdateConfig, jobVMwareSourceEndpoint, vmWareJobName),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckJobVmwareUpdated(),
				),
			},
		},
	})
}

func testAccCheckJobVmwareDestroy(s *terraform.State) error {
	var cohesityConfig = testAccProvider.Meta().(Config)
	client, err := CohesityManagementSdk.NewCohesitySdkClient(cohesityConfig.clusterVip,
		cohesityConfig.clusterUsername, cohesityConfig.clusterPassword, cohesityConfig.clusterDomain)
	if err != nil {
		log.Printf(err.Error())
		return fmt.Errorf("Failed to authenticate with Cohesity")
	}

	resourceName := "cohesity_job_vmware.terraform_vmware_job"
	resource, ok := s.RootModule().Resources[resourceName]
	if !ok {
		return fmt.Errorf("Can't find the resource: %s", resourceName)
	}
	if resource.Primary.ID == "" {
		return fmt.Errorf("The ID for resoource %s for is not set", resourceName)
	}
	//parse a decimal string of base 10 and converts into int64
	jobID, _ := strconv.ParseInt(resource.Primary.ID, 10, 64)
	result, err := client.ProtectionJobs().GetProtectionJobs([]int64{jobID},
		nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
	if err != nil || len(result) == 0 {
		return nil
	}
	return fmt.Errorf("Failed to delete the VMware protection job")
}

func testAccCheckJobVmwareCreated() resource.TestCheckFunc {
	return func(s *terraform.State) error {
		var cohesityConfig = testAccProvider.Meta().(Config)
		client, err := CohesityManagementSdk.NewCohesitySdkClient(cohesityConfig.clusterVip,
			cohesityConfig.clusterUsername, cohesityConfig.clusterPassword, cohesityConfig.clusterDomain)
		if err != nil {
			log.Printf(err.Error())
			return errors.New("Failed to authenticate with Cohesity")
		}
		resourceName := "cohesity_job_vmware.terraform_vmware_job"
		resource, ok := s.RootModule().Resources[resourceName]
		if !ok {
			return fmt.Errorf("Can't find the resource: %s", resourceName)
		}
		if resource.Primary.ID == "" {
			return fmt.Errorf("The ID for resource %s is not set", resourceName)
		}
		//parse a decimal string of base 10 and converts into int64
		jobID, _ := strconv.ParseInt(resource.Primary.ID, 10, 64)
		result, err := client.ProtectionJobs().GetProtectionJobs([]int64{jobID},
			nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
		if err != nil || len(result) == 0 {
			return fmt.Errorf("Failed to create VMware protection job")
		}
		//validate protection job parameters with values from testAccJobVmwareCreateConfig
		//configuration
		if result[0].Name != vmWareJobName || *result[0].FullProtectionSlaTimeMins != 200 ||
			*result[0].IncrementalProtectionSlaTimeMins != 140 {
			return fmt.Errorf("Failed to validate the created VMware protection job")
		}
		return nil
	}
}

func testAccCheckJobVmwareUpdated() resource.TestCheckFunc {
	return func(s *terraform.State) error {
		var cohesityConfig = testAccProvider.Meta().(Config)
		client, err := CohesityManagementSdk.NewCohesitySdkClient(cohesityConfig.clusterVip,
			cohesityConfig.clusterUsername, cohesityConfig.clusterPassword, cohesityConfig.clusterDomain)
		if err != nil {
			log.Printf(err.Error())
			return errors.New("Failed to authenticate with Cohesity")
		}
		resourceName := "cohesity_job_vmware.terraform_vmware_job"
		resource, ok := s.RootModule().Resources[resourceName]
		if !ok {
			return fmt.Errorf("Can't find the resource: %s", resourceName)
		}
		if resource.Primary.ID == "" {
			return fmt.Errorf("The ID for resoource %s for is not set", resourceName)
		}
		//parse a decimal string of base 10 and converts into int64
		jobID, _ := strconv.ParseInt(resource.Primary.ID, 10, 64)
		result, err := client.ProtectionJobs().GetProtectionJobs([]int64{jobID},
			nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
		if err != nil || len(result) == 0 {
			return fmt.Errorf("Failed to find the VMware protection job")
		}
		//validate protection job parameters with values from testAccJobVmwareUpdateConfig
		//configuration
		if result[0].Name != vmWareJobName || *result[0].FullProtectionSlaTimeMins != 180 ||
			*result[0].IncrementalProtectionSlaTimeMins != 120 {
			return fmt.Errorf("Failed to validate the updated VMware protection job")
		}
		return nil
	}
}

const testAccJobVmwareCreateConfig = `
resource "cohesity_source_vmware" "vmware_source"{
	endpoint = "%s"
	username = "administrator"
	vmware_type = "VCenter"
	cap_streams_per_datastore = true
	number_of_streams = 5
	enable_latency_throttling = true
	new_task_latency = 110
	active_task_latency = 120
}

resource "cohesity_job_vmware" "terraform_vmware_job" {
	name = "%s"
	protection_source = cohesity_source_vmware.vmware_source.endpoint
	policy = "Bronze"
	storage_domain = "DefaultStorageDomain"
	delete_snapshots = true
	full_sla = 200
	incremental_sla = 140
	qos_type = "BackupSSD"
	priority = "Low"
}
`

const testAccJobVmwareUpdateConfig = `
resource "cohesity_source_vmware" "vmware_source"{
	endpoint = "%s"
	username = "administrator"
	vmware_type = "VCenter"
	cap_streams_per_datastore = true
	number_of_streams = 5
	enable_latency_throttling = true
	new_task_latency = 110
	active_task_latency = 120
}

resource "cohesity_job_vmware" "terraform_vmware_job" {
	name = "%s"
	protection_source = cohesity_source_vmware.vmware_source.endpoint
	policy = "Bronze"
	storage_domain = "DefaultStorageDomain"
	delete_snapshots = true
	full_sla = 180
	incremental_sla = 120
	qos_type = "BackupSSD"
	priority = "Low"
}
`


cohesity/resource_cohesity_job_vmware.go
package cohesity

import (
	"errors"
	"log"
	"strconv"

	CohesityManagementSdk "github.com/cohesity/management-sdk-go/managementsdk"
	"github.com/cohesity/management-sdk-go/models"
	"github.com/golang-collections/collections/queue"
	"github.com/hashicorp/terraform/helper/schema"
)

func resourceCohesityJobVMware() *schema.Resource {
	return &schema.Resource{
		Create: resourceCohesityJobVMwareCreate,
		Read:   resourceCohesityJobVMwareRead,
		Delete: resourceCohesityJobVMwareDelete,
		Update: resourceCohesityJobVMwareUpdate,
		Schema: map[string]*schema.Schema{
			"name": {
				Type:        schema.TypeString,
				Required:    true,
				Description: "Specifies the name of the Protection Job",
			},
			"protection_source": {
				Type:        schema.TypeString,
				Required:    true,
				Description: "Specifies the Protection source",
			},
			"timezone": {
				Type:        schema.TypeString,
				Optional:    true,
				Default:     "America/Los_Angeles",
				Description: "Specifies the timezone in the format: Area/Location",
			},
			"include": {
				Type:     schema.TypeSet,
				Optional: true,
				Elem:     &schema.Schema{Type: schema.TypeString},
				Set:      schema.HashString,
				Description: `Specifies the list of vm's from the
				protection source to be protected by this protection job`,
			},
			"exclude": {
				Type:     schema.TypeSet,
				Optional: true,
				Elem:     &schema.Schema{Type: schema.TypeString},
				Set:      schema.HashString,
				Description: `Specifies the list of vm's from the protection source
				that should not be protected and are excluded from being
				backed up by the protection job`,
			},
			"policy": {
				Type:        schema.TypeString,
				Required:    true,
				Description: "Specifies the protection policy",
			},
			"storage_domain": {
				Type:     schema.TypeString,
				Required: true,
				Description: `Specifies the storage domain where this job
				writes data`,
			},
			"qos_type": {
				Type:     schema.TypeString,
				Optional: true,
				Default:  "BackupHDD",
				Description: `Specifies the QoS policy type to use for this
				Protection Job. 'BackupHDD' indicates the Cohesity Cluster
				writes data directly to the HDD tier for this Protection Job.
				This is the recommended setting. 'BackupSSD' indicates the
				Cohesity Cluster writes data directly to the SSD tier for this
				Protection Job. Only specify this policy if you need fast ingest
				speed for a small number of Protection Jobs`,
			},
			"full_sla": {
				Type:     schema.TypeInt,
				Optional: true,
				Default:  120,
				Description: `Specifies the number of minutes that a Job Run
				of a Full (no CBT) backup schedule is expected to complete,
				which is known as a Service-Level Agreement (SLA).A SLA
				violation is reported when the run time of a Job Run exceeds
				the SLA time period specified for this backup schedule.`,
			},
			"incremental_sla": {
				Type:     schema.TypeInt,
				Optional: true,
				Default:  60,
				Description: `Specifies the number of minutes that a Job Run
				of a CBT-based backup schedule is expected to complete, which
				is known as a Service-Level Agreement (SLA). A SLA violation
				is reported when the run time of a Job Run exceeds the SLA
				time period specified for this backup schedule.`,
			},
			"priority": {
				Type:     schema.TypeString,
				Optional: true,
				Default:  "Medium",
				Description: `Specifies the priority of execution for a
				Protection Job`,
			},
			"delete_snapshots": {
				Type:     schema.TypeBool,
				Optional: true,
				Default:  false,
				Description: `Specifies if snapshots generated by the Protection
				Job should also be deleted when the Job is deleted.`,
			},
		},
	}
}

//empty structure used as vaule in Map, this is used
//to mimic HashSet with Map in golang
type empty struct{}

//parse the protection source node structure for vm ids
func parseGetSourceIDs(sources map[string]empty, sourceNode []interface{}) []int64 {
	var nodesQueue *queue.Queue
	nodesQueue = queue.New()
	var sourceIDs = []int64{}
	for _, node := range sourceNode {
		nodesQueue.Enqueue(node)
	}

	for nodesQueue.Len() != 0 && len(sources) != 0 {
		node := nodesQueue.Dequeue()
		var name = (((node.(map[string]interface {
		}))["protectionSource"]).(map[string]interface{}))["name"].(string)
		if _, ok := sources[name]; ok {
			id := int64((((node.(map[string]interface {
			}))["protectionSource"]).(map[string]interface{}))["id"].(float64))
			sourceIDs = append(sourceIDs, id)
			delete(sources, name)
			continue
		}
		nodes, ok := node.(map[string]interface{})["nodes"]
		if ok == true && len(nodes.([]interface{})) != 0 {
			for _, node := range nodes.([]interface{}) {
				nodesQueue.Enqueue(node)
			}
		}
	}
	return sourceIDs
}

func resourceCohesityJobVMwareCreate(resourceData *schema.ResourceData, configMetaData interface{}) error {
	var cohesityConfig = configMetaData.(Config)
	// authenticate with Cohesity cluster
	client, err := CohesityManagementSdk.NewCohesitySdkClient(cohesityConfig.clusterVip,
		cohesityConfig.clusterUsername, cohesityConfig.clusterPassword, cohesityConfig.clusterDomain)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to authenticate with Cohesity")
	}
	var requestBody models.ProtectionJobRequestBody
	var name = resourceData.Get("name").(string)
	var protectionSource = resourceData.Get("protection_source").(string)
	var fullSLA = int64(resourceData.Get("full_sla").(int))
	var incrementalSLA = int64(resourceData.Get("incremental_sla").(int))
	var timezone = resourceData.Get("timezone").(string)
	var protectionSourceID int64

	//get the policy id
	log.Printf("[INFO] Get the protection policy id")
	policy, err := client.ProtectionPolicies().GetProtectionPolicies(nil,
		[]string{resourceData.Get("policy").(string)}, nil, nil, nil, nil)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Error in getting the protection policy")
	} else if err == nil && len(policy) == 0 {
		return errors.New("Failed to find the protection policy on Cohesity cluster")
	}
	policyID := *policy[0].Id

	//get the storage domain id
	log.Printf("[INFO] Get the storage domain id")
	storageDomain, err := client.ViewBoxes().GetViewBoxes(nil, nil, nil,
		[]string{resourceData.Get("storage_domain").(string)}, nil, nil, nil)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Error in getting the storage domain")
	} else if err == nil && len(storageDomain) == 0 {
		return errors.New("Failed to find the storage domain on Cohesity cluster")
	}
	storageDomainID := *storageDomain[0].Id

	//get the protection source id
	log.Printf("[INFO] Get the protection source id")
	sources, err := client.ProtectionSources().ListProtectionSourcesRootNodes(nil,
		[]models.EnvironmentListProtectionSourcesRootNodesEnum{models.
			EnvironmentListProtectionSourcesRootNodes_KVMWARE}, nil)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Error in getting the protection source")
	}
	for _, source := range sources {
		if *source.ProtectionSource.Name == protectionSource {
			protectionSourceID = *source.ProtectionSource.Id
		}
	}
	if protectionSourceID == 0 {
		return errors.New("Failed to find the protection source on Cohesity cluster")
	}

	//backup specified vm's or enter vCenter
	protectionSourceWithClildObjects, err := client.ProtectionSources().
		ListProtectionSources(&protectionSourceID, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to get the protection source")
	}
	log.Printf("[INFO] Set the source id's for the protection job")
	if len(resourceData.Get("include").(*schema.Set).List()) == 0 {
		var sourceIds = []int64{protectionSourceID}
		requestBody.SourceIds = &sourceIds
	} else {
		var emptyValue empty
		includeSourceSet := make(map[string]empty)
		for _, source := range resourceData.Get("include").(*schema.Set).List() {
			includeSourceSet[source.(string)] = emptyValue
		}
		n := (*protectionSourceWithClildObjects[0].
			Nodes)[0].(map[string]interface{})["nodes"].([]interface{})
		includeSourceIDs := parseGetSourceIDs(includeSourceSet, n)
		if len(resourceData.Get("include").(*schema.Set).List()) != len(includeSourceIDs) {
			return errors.New("Failed to find all the included protection source objects")
		}
		requestBody.SourceIds = &includeSourceIDs
	}

	//exclude the specified vm's
	if len(resourceData.Get("exclude").(*schema.Set).List()) != 0 {
		log.Printf("[INFO] Set the exclude source id's for the protection job")
		var emptyValue empty
		excludeSourceSet := make(map[string]empty)
		for _, source := range resourceData.Get("exclude").(*schema.Set).List() {
			excludeSourceSet[source.(string)] = emptyValue
		}
		n := (*protectionSourceWithClildObjects[0].
			Nodes)[0].(map[string]interface{})["nodes"].([]interface{})
		excludeSourceIDs := parseGetSourceIDs(excludeSourceSet, n)
		if len(resourceData.Get("exclude").(*schema.Set).List()) != len(excludeSourceIDs) {
			return errors.New("Failed to find all the excluded protection source objects")
		}
		requestBody.ExcludeSourceIds = &excludeSourceIDs
	}

	requestBody.Name = name
	requestBody.ParentSourceId = &protectionSourceID
	requestBody.PolicyId = policyID
	requestBody.ViewBoxId = storageDomainID
	requestBody.Timezone = &timezone
	requestBody.Environment = models.EnvironmentProtectionJobRequestBody_KVMWARE

	//set the qos type
	log.Printf("[INFO] Set the qos type for the protection job")
	switch resourceData.Get("qos_type").(string) {
	case "BackupHDD":
		requestBody.QosType = models.QosType_KBACKUPHDD
	case "BackupSSD":
		requestBody.QosType = models.QosType_KBACKUPSSD
	}

	//set the priority
	log.Printf("[INFO] Get the priority for the protection job")
	switch resourceData.Get("priority").(string) {
	case "Medium":
		requestBody.Priority = models.Priority_KMEDIUM
	case "Low":
		requestBody.Priority = models.Priority_KLOW
	case "High":
		requestBody.Priority = models.Priority_KHIGH
	}

	//set the SLA
	log.Printf("[INFO] Get the SLA for the protection job")
	requestBody.FullProtectionSlaTimeMins = &fullSLA
	requestBody.IncrementalProtectionSlaTimeMins = &incrementalSLA

	result, err := client.ProtectionJobs().CreateProtectionJob(&requestBody)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to create the protection job")
	}
	resourceData.SetId(strconv.FormatInt(*result.Id, 10))
	log.Printf("[INFO] Successfully created the protection job %s", name)
	return resourceCohesityJobVMwareRead(resourceData, configMetaData)
}

func resourceCohesityJobVMwareRead(resourceData *schema.ResourceData, configMetaData interface{}) error {
	var cohesityConfig = configMetaData.(Config)
	client, err := CohesityManagementSdk.NewCohesitySdkClient(cohesityConfig.clusterVip,
		cohesityConfig.clusterUsername, cohesityConfig.clusterPassword, cohesityConfig.clusterDomain)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to authenticate with Cohesity")
	}
	//parse a decimal string of base 10 and converts into int64
	sourceID, _ := strconv.ParseInt(resourceData.Id(), 10, 64)

	log.Printf("[INFO] Read Cohesity VMware protection job")
	result, err := client.ProtectionJobs().GetProtectionJobs([]int64{sourceID},
		nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
	if err != nil || len(result) == 0 {
		resourceData.SetId("")
	}
	return nil
}

func resourceCohesityJobVMwareDelete(resourceData *schema.ResourceData, configMetaData interface{}) error {
	var cohesityConfig = configMetaData.(Config)
	client, err := CohesityManagementSdk.NewCohesitySdkClient(cohesityConfig.clusterVip,
		cohesityConfig.clusterUsername, cohesityConfig.clusterPassword, cohesityConfig.clusterDomain)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to authenticate with Cohesity")
	}

	//parse a decimal string of base 10 and converts into int64
	jobID, _ := strconv.ParseInt(resourceData.Id(), 10, 64)
	log.Printf("[INFO] Delete VMware protection job %s", resourceData.
		Get("name").(string))
	deleteSnapshots := resourceData.Get("delete_snapshots").(bool)
	deleteProtectionJobParam := models.
		DeleteProtectionJobParam{DeleteSnapshots: &deleteSnapshots}
	err = client.ProtectionJobs().DeleteProtectionJob(jobID, &deleteProtectionJobParam)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to delete the protection job")
	}
	return nil
}

func resourceCohesityJobVMwareUpdate(resourceData *schema.ResourceData, configMetaData interface{}) error {
	resourceData.Partial(true)
	var cohesityConfig = configMetaData.(Config)
	client, err := CohesityManagementSdk.NewCohesitySdkClient(cohesityConfig.clusterVip,
		cohesityConfig.clusterUsername, cohesityConfig.clusterPassword, cohesityConfig.clusterDomain)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to authenticate with Cohesity")
	}
	log.Printf("[INFO] Update VMware protection job")
	var requestBody models.ProtectionJobRequestBody

	//parse a decimal string of base 10 and converts into int64
	jobID, _ := strconv.ParseInt(resourceData.Id(), 10, 64)
	result, err := client.ProtectionJobs().GetProtectionJobs([]int64{jobID},
		nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)

	var name = resourceData.Get("name").(string)
	var timezone = resourceData.Get("timezone").(string)
	var fullSLA = int64(resourceData.Get("full_sla").(int))
	var incrementalSLA = int64(resourceData.Get("incremental_sla").(int))

	requestBody.Name = name
	requestBody.Timezone = &timezone
	requestBody.FullProtectionSlaTimeMins = &fullSLA
	requestBody.IncrementalProtectionSlaTimeMins = &incrementalSLA

	if resourceData.HasChange("protection_source") {
		return errors.New("Can't update the protection source in VMware protection job")
	}
	requestBody.ParentSourceId = result[0].ParentSourceId

	switch resourceData.Get("qos_type").(string) {
	case "BackupHDD":
		requestBody.QosType = models.QosType_KBACKUPHDD
	case "BackupSSD":
		requestBody.QosType = models.QosType_KBACKUPSSD
	}

	switch resourceData.Get("priority").(string) {
	case "Medium":
		requestBody.Priority = models.Priority_KMEDIUM
	case "Low":
		requestBody.Priority = models.Priority_KLOW
	case "High":
		requestBody.Priority = models.Priority_KHIGH
	}

	if resourceData.HasChange("policy") {
		log.Printf("[INFO] Update VMware protection job policy")
		policy, err := client.ProtectionPolicies().GetProtectionPolicies(nil,
			[]string{resourceData.Get("policy").(string)}, nil, nil, nil, nil)
		if err != nil {
			log.Printf(err.Error())
			return errors.New("Error in getting the protection policy")
		} else if err == nil && len(policy) == 0 {
			return errors.New("Failed to find the protection policy on Cohesity cluster")
		}
		policyID := *policy[0].Id
		requestBody.PolicyId = policyID

	} else {
		requestBody.PolicyId = result[0].PolicyId
	}

	if resourceData.HasChange("storage_domain") {
		log.Printf("[INFO] Update VMware protection job storage domain")
		storageDomain, err := client.ViewBoxes().GetViewBoxes(nil, nil, nil,
			[]string{resourceData.Get("storage_domain").(string)}, nil, nil, nil)
		if err != nil {
			log.Printf(err.Error())
			return errors.New("Error in getting the storage domain")
		} else if err == nil && len(storageDomain) == 0 {
			return errors.New("Failed to find the storage domain on Cohesity cluster")
		}
		storageDomainID := *storageDomain[0].Id
		requestBody.ViewBoxId = storageDomainID

	} else {
		requestBody.ViewBoxId = result[0].ViewBoxId
	}
	protectionSourceWithClildObjects, err := client.ProtectionSources().ListProtectionSources(result[0].
		ParentSourceId, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to get the protection source")
	}

	if resourceData.HasChange("include") {
		log.Printf("[INFO] Update VMware protection job sources id's")
		if len(resourceData.Get("include").(*schema.Set).List()) == 0 {
			var sourceIds = []int64{*result[0].ParentSourceId}
			requestBody.SourceIds = &sourceIds
		} else {
			var emptyValue empty
			includeSourceSet := make(map[string]empty)
			for _, source := range resourceData.Get("include").(*schema.Set).List() {
				includeSourceSet[source.(string)] = emptyValue
			}
			n := (*protectionSourceWithClildObjects[0].
				Nodes)[0].(map[string]interface{})["nodes"].([]interface{})
			includeSourceIDs := parseGetSourceIDs(includeSourceSet, n)
			if len(resourceData.Get("include").(*schema.Set).List()) != len(includeSourceIDs) {
				return errors.New("Failed to find all the included protection source objects")
			}
			requestBody.SourceIds = &includeSourceIDs
		}
	} else {
		requestBody.SourceIds = result[0].SourceIds
	}

	if resourceData.HasChange("exclude") {
		log.Printf("[INFO] Update VMware protection job exclude source id's")
		if len(resourceData.Get("exclude").(*schema.Set).List()) != 0 {
			var emptyValue empty
			excludeSourceSet := make(map[string]empty)
			for _, source := range resourceData.Get("exclude").(*schema.Set).List() {
				excludeSourceSet[source.(string)] = emptyValue
			}
			n := (*protectionSourceWithClildObjects[0].
				Nodes)[0].(map[string]interface{})["nodes"].([]interface{})
			excludeSourceIDs := parseGetSourceIDs(excludeSourceSet, n)
			if len(resourceData.Get("exclude").(*schema.Set).List()) != len(excludeSourceIDs) {
				return errors.New("Failed to find all the excluded protection source objects")
			}
			requestBody.ExcludeSourceIds = &excludeSourceIDs
		} else {
			requestBody.ExcludeSourceIds = nil
		}
	} else {
		requestBody.ExcludeSourceIds = result[0].ExcludeSourceIds
	}
	_, err = client.ProtectionJobs().UpdateProtectionJob(&requestBody, *result[0].Id)
	if err != nil {
		log.Printf(err.Error())
		return errors.New("Failed to update the VMware protection job")
	}
	resourceData.Partial(false)
	return nil
}
